scylla-cql/src/types/serialize/writers.rs

```diff
@@ -174,11 +302,19 @@ impl<'buf> CellValueBuilder<'buf> {
     /// CQL cell size (which is i32::MAX).
     #[inline]
     pub fn finish(self) -> Result<WrittenCellProof<'buf>, CellOverflowError> {
-        let value_len: i32 = (self.buf.len() - self.starting_pos - 4)
-            .try_into()
-            .map_err(|_| CellOverflowError)?;
-        self.buf[self.starting_pos..self.starting_pos + 4]
-            .copy_from_slice(&value_len.to_be_bytes());
+        if let Some(buffer) = self.variable_length_buffer {
+            let value_len = buffer.len();
+            let mut len = Vec::new();
+            types::unsigned_vint_encode(value_len as u64, &mut len);
+            self.buf.extend_from_slice(&len);
+            self.buf.extend_from_slice(&buffer);
+        } else {
+            let value_len: i32 = (self.buf.len() - self.starting_pos - 4)
+                .try_into()
+                .map_err(|_| CellOverflowError)?;
+            self.buf[self.starting_pos..self.starting_pos + 4]
+                .copy_from_slice(&value_len.to_be_bytes());
+        }
```

Comment:
:thought_balloon: The logic around `Cell*` got much convoluted with these alterations. Let's think how we can make it more digestable. @Lorak-mmk @muzarski